File: Data Structures & Algorithms/queue/submission-0.cs
class Deque {

    ListNode head;
    ListNode tail;
    int count;

    public Deque() {
        head = null;
        tail = null;
        count = 0;
    }

    public bool isEmpty() {
        return count == 0 ? true : false;
    }

    public void append(int value) {
        ListNode node = new ListNode(value);
        if(tail == null) {
            head = node;
            tail = node;
        }
        else {
            tail.next = node;
            node.prev = tail;
            tail = node;
        }
        count++;
    }

    public void appendleft(int value) {
        ListNode node = new ListNode(value);
        if(head == null) {
            head = node;
            tail = node;
        }
        else {
            node.next = head;
            head.prev = node;
            head = node;
        }
        count++;
    }

    public int pop() {
        if(count == 0) return -1;
        if(head == tail) {
            ListNode node = new ListNode(0);
            node = head;
            head = null;
            tail = null;
            count = 0;
            return node.value;
        }
        else {
            ListNode node = new ListNode(0);
            node = tail;
            tail = tail.prev;
            tail.next = null;
            count--;
            return node.value;
        }
    }

    public int popleft() {
        if(count == 0) return -1;
        if(head == tail) {
            ListNode node = new ListNode(0);
            node = head;
            head = null;
            tail = null;
            count = 0;
            return node.value;
        }
        else {
            ListNode node = new ListNode(0);
            node = head;
            head = head.next;
            head.prev = null;
            count--;
            return node.value;
        }
    }
}

public class ListNode {
    public int value;
    public ListNode prev;
    public ListNode next;

    public ListNode(int value) {
        this.value = value;
        this.prev = null;
        this.next = null;
    }
}

File: Data Structures & Algorithms/queue/submission-0.py
class ListNode:
    def __init__(self, val = 0, prev = None, next = None):
        self.val = val
        self.prev = prev
        self.next = next

class Deque:
    
    def __init__(self):
        self.head = ListNode()
        self.tail = ListNode()
        self.nodeCount = 0
        self.head.next, self.tail.prev = self.tail, self.head 

    def isEmpty(self) -> bool:
        return True if self.nodeCount == 0 else False

    def append(self, value: int) -> None:
        newNode = ListNode(value, self.tail.prev, self.tail)
        self.tail.prev.next, self.tail.prev = newNode, newNode
        self.nodeCount+=1

    def appendleft(self, value: int) -> None:
        newNode = ListNode(value, self.head, self.head.next)
        self.head.next.prev, self.head.next = newNode, newNode
        self.nodeCount+=1

    def pop(self) -> int:
        if self.isEmpty():
            return -1
        
        toBeDeletedNode = self.tail.prev
        newLeftNode = self.tail.prev.prev
        
        newLeftNode.next, self.tail.prev = self.tail, newLeftNode

        self.nodeCount-=1
        return toBeDeletedNode.val

    def popleft(self) -> int:
        if self.isEmpty():
            return -1
        
        toBeDeletedNode = self.head.next
        newRightNode = self.head.next.next
        newRightNode.prev, self.head.next = self.head, newRightNode

        self.nodeCount-=1
        return toBeDeletedNode.val
